### DISCUSSION TITLE
Generic Combine HL7 Segments

### DISCUSSION THREAD

**Discussion by jtyme** [nextgenhealthcare/connect#5699] [Q&A] [answered]
I know I can do the following code, but is it possible to make it generic where you can pass in the segment value you want to combine on?
`var mdmObxLength = xmlMdmFullRecord..OBX.length();
 var newIndex = mdmObxLength + 1;

//loop through the ORU and get each OBX
for each(obx in xmlOruFullRecord..OBX){
  obx['OBX.1']['OBX.1.1'] = newIndex;

//add the ORU Obx Segment to the MDM.
   xmlMdmFullRecord.appendChild(obx);
   newIndex++;
}`

I have tried 
`function genericCombine(parentMessage, childMessage, seg){
 var parentSegLength = parentMessage..seg.length();
var newIndex = parentSegLength + 1;
for each(s in childMessage..seg){
   s[seg + '.1'][seg + '.1.1'] = newIndex;
   parentMessage.appendChild(s);
   newIndex++;
}
}
`
When I run the above I error of Cannot call method "length" of undefined
my test call is mdmFullMessage = genericCombine(mdmFullMessage, oruFullMessage, 'OBX');
I am assuming because seg is passed in as a string 'OBX' it is seeing it in this fashion parentMessage..'OBX'.length()

**Comment by tonygermano** [comment, marked as answer]
First, when dealing with xml `..` is the `descendants` operator, and `.` is the `children` operator. You almost always want to use `.` when working with hl7. Using bracket notation always only looks for children.

I didn't get any errors when I tried your code, but it also didn't do anything. Be aware that this function alters both the parentMessage and the childMessage (when renumbering the first field.)

```js
function genericCombine(parentMessage, childMessage, seg) {
  // The following is looking for any descendant of parentMessage named "seg."
  // There aren't any, so the length should be 0.
  // var parentSegLength = parentMessage..seg.length();
  var parentSegLength = parentMessage[seg].length();
  var newIndex = parentSegLength + 1;
  // This needs to change to bracket notation, too. Also, s should be scoped to this function instead of global.
  // for each(s in childMessage..seg) {
  for each(var s in childMessage[seg]) {
    s[seg + '.1'][seg + '.1.1'] = newIndex;
    parentMessage.appendChild(s);
    newIndex++;
  }
}
```
